dfget/core/downloader/p2p_downloader.go

```diff
@@ -203,18 +203,45 @@ func (p2p *P2PDownloader) pullPieceTask(item *Piece) (
 	return res, err
 }
 
-func (p2p *P2PDownloader) pullRate(data *types.PullPieceTaskResponseContinueData) {
+// getPullRate get download rate limit dynamically.
+func (p2p *P2PDownloader) getPullRate(data *types.PullPieceTaskResponseContinueData) int {
+	var localRate int
 
+	if p2p.Cfg.LocalLimit > 0 {
+		localRate = p2p.Cfg.LocalLimit
+	} else {
+		localRate = data.DownLink * 1024
+	}
+
+	// Calculate the download speed limit
+	// that the current download task can be assigned
+	// by the uploader server.
+	// TODO: Reduce the frequency of requests.
+	url := fmt.Sprintf("http://%s:%d%s%s", p2p.Cfg.RV.LocalIP, p2p.Cfg.RV.PeerPort, config.LocalHTTPPathRate, p2p.taskFileName)
+	headers := make(map[string]string)
+	headers["rateLimit"] = strconv.Itoa(localRate)
+	resp, err := util.Do(url, headers, util.DefaultTimeout)
+	if err != nil {
+		p2p.Cfg.ClientLogger.Errorf("failed to pullRate: %v", err)
+		return localRate
+	}
+
+	reqRate, err := strconv.Atoi(resp)
+	if err != nil {
+		return localRate
+	}
+	return reqRate
 }
 
-func (p2p *P2PDownloader) startTask(data *types.PullPieceTaskResponseContinueData) {
+func (p2p *P2PDownloader) startTask(data *types.PullPieceTaskResponseContinueData, rateLimit int) {
 	powerClient := &PowerClient{
 		taskID:      p2p.taskID,
 		node:        p2p.node,
 		pieceTask:   data,
 		cfg:         p2p.Cfg,
 		queue:       p2p.queue,
 		clientQueue: p2p.clientQueue,
+		rateLimit:   rateLimit,
 	}
 	powerClient.Run()
```

Comment:
How about does `PowerClient.Run` return `Piece` and `error`? Then we don't need to pass `queue` and `clientQueue` to `PowerClient`. We can process them in this function according to the result of `PowerClient.Run`.

Reply:
Maybe it's not a good idea. It is not always necessary to put the returned piece both into `quene` and `clientQueue`.